raft.go

```diff
@@ -1468,3 +1632,71 @@ func (r *Raft) setState(state RaftState) {
 		r.observe(state)
 	}
 }
+
+// LookupServer looks up a server by ServerID.
+func (r *Raft) lookupServer(id ServerID) *Server {
+	for _, server := range r.configurations.latest.Servers {
+		if server.ID != r.localID {
+			return &server
+		}
+	}
+	return nil
+}
+
+// pickServer returns the follower that is most up to date.
+func (r *Raft) pickServer() *Server {
+	target := r.getLastIndex()
+	var pick *Server
+	var current uint64
+	for _, server := range r.configurations.latest.Servers {
+		if server.ID == r.localID {
+			continue
+		}
+		state, ok := r.leaderState.replState[server.ID]
+		if !ok {
+			continue
+		}
+		// return early if this server is up to date
+		if state.nextIndex > target {
+			return &server
+		}
```

Comment:
Is this necessary?

The loop isn't long and would always find the most up-to-date anyway so I don't think it's a big performance win. In some cases it might even work out to be less-than-optimal for transfer time because there may have been 10 new logs recorded since our `getLastIndex` above and the first server is all 10 behind while another server in the list might be only 1 behind.

Reply:
Agreed. Fixed in https://github.com/hashicorp/raft/pull/306/commits/8deca55fc7984ab0a0dd85b9ab29bccdc6711b6e.